services/api/src/api/routes/endpoint.py

```diff
@@ -28,21 +28,171 @@
     """Definition of supported endpoints and its relation with processing steps."""
 
-    processing_steps_by_endpoint: Mapping[str, List[ProcessingStep]]
+    steps_by_input_type_and_endpoint: StepsByInputTypeAndEndpoint
 
     def __init__(self, graph: ProcessingGraph, endpoint_config: EndpointConfig):
-        self.definition = {
-            endpoint: [graph.get_step(step) for step in processing_steps]
-            for endpoint, processing_steps in endpoint_config.specification.items()
+        self.steps_by_input_type_and_endpoint = {
+            endpoint: {
+                input_type: [graph.get_step(step_name) for step_name in step_names]
+                for input_type, step_names in step_names_by_input_type.items()
+            }
+            for endpoint, step_names_by_input_type in endpoint_config.step_names_by_input_type_and_endpoint.items()
         }
 
 
-def create_endpoint(
+def get_cache_entry_from_steps(
     processing_steps: List[ProcessingStep],
+    dataset: str,
+    config: Optional[str],
+    split: Optional[str],
+    init_processing_steps: List[ProcessingStep],
+    hf_endpoint: str,
+    hf_token: Optional[str] = None,
+) -> CacheEntry:
+    """Gets the cache from the first successful step in the processing steps list.
+    If no successful result is found, it will return the last one even if it's an error,
+    Checks if job is still in progress by each processing step in case of no entry found.
+    Raises:
+        ResponseNotReadyError: if no result is found.
+
+    Returns: the cached record
+    """
+    last_result = None
+    for processing_step in processing_steps:
+        try:
+            last_result = get_response(
+                kind=processing_step.cache_kind,
+                dataset=dataset,
+                config=config,
+                split=split,
+            )
+
+            if last_result["http_status"] == HTTPStatus.OK:
+                return last_result
+        except DoesNotExist:
+            logging.debug(
+                f"processing_step={processing_step.name} dataset={dataset} "
+                f"config={config} split={split} no entry found"
+            )
+            try:
+                check_in_process(
+                    processing_step=processing_step,
+                    init_processing_steps=init_processing_steps,
+                    dataset=dataset,
+                    config=config,
+                    split=split,
+                    hf_endpoint=hf_endpoint,
+                    hf_token=hf_token,
+                )
+            except (PreviousStepError, DatasetError):
+                raise ResponseNotFoundError("Not found.")
+    if last_result:
+        return last_result
+
+    raise ResponseNotReadyError(
+        "The server is busier than usual and the response is not ready yet. Please retry later."
+    )
+
+
+class InputTypeValidator(ABC):
+    def __init__(self, input_type: InputType) -> None:
+        self.input_type = input_type
```

Comment:
Nice (the code is longer, but the logic is good)

nit: input_type should be an abstract field, set directly in the children classes, not a parameter (it's fixed for a given class)

Reply:
Makes sense, thanks!